pallets/identity/src/keys.rs

```diff
@@ -214,36 +230,83 @@ impl<T: Config> Module<T> {
 
+        // Replace primary key of the owner that initiated key rotation.
+        record.primary_key = new_primary_key.clone();
+        Self::deposit_event(RawEvent::PrimaryKeyUpdated(
+            target_did,
+            old_primary_key.clone(),
+            new_primary_key,
+        ));
+        DidRecords::<T>::insert(target_did, record);
+        Ok(())
+    }
+
+    /// Processes primary key rotation.
+    pub fn rotate_primary_key_and_remove(
+        sender: T::AccountId,
+        rotation_for_did: IdentityId,
+        optional_cdd_auth_id: Option<u64>,
+    ) -> DispatchResult {
+        let old_primary_key = <DidRecords<T>>::get(rotation_for_did).primary_key;
         // Ensure that it is safe to unlink the primary key from the did.
         Self::ensure_key_unlinkable_from_did(&old_primary_key)?;
 
-        // Replace primary key of the owner that initiated key rotation.
-        Self::unlink_account_key_from_did(&old_primary_key, rotation_for_did);
-        record.primary_key = sender.clone();
-        Self::link_account_key_to_did(&sender, rotation_for_did);
-        <DidRecords<T>>::insert(&rotation_for_did, record);
+        Self::common_rotate_primary_key(rotation_for_did, sender, optional_cdd_auth_id)?;
 
-        Self::deposit_event(RawEvent::PrimaryKeyUpdated(
-            rotation_for_did,
-            old_primary_key,
-            sender,
-        ));
+        Self::unlink_account_key_from_did(&old_primary_key, rotation_for_did);
         Ok(())
     }
 
+    /// Accepts a primary key rotation.
+    /// Differs from accept_primary_key_rotation in that it will leave the old primary key as a
+    /// secondary key with the permissions specified in the corresponding RotatePrimaryKeyToSecondary authorization
+    /// instead of unlinking the primary key.
+    crate fn base_rotate_primary_key_to_secondary(
+        origin: T::Origin,
+        rotation_auth_id: u64,
+        optional_cdd_auth_id: Option<u64>,
+    ) -> DispatchResult {
+        let new_primary_key = ensure_signed(origin)?;
+        let new_primary_key_signer = Signatory::Account(new_primary_key.clone());
+        Self::accept_auth_with(
+            &new_primary_key_signer,
+            rotation_auth_id,
+            |data, target_did| {
+                let old_primary_key = <DidRecords<T>>::get(target_did).primary_key;
+
+                Self::common_rotate_primary_key(target_did, new_primary_key, optional_cdd_auth_id)?;
+
+                let perms = extract_auth!(data, RotatePrimaryKeyToSecondary(p));
+
+                // Link the secondary key.
+                <DidRecords<T>>::mutate(target_did, |record| {
+                    let sk = SecondaryKey::new(Signatory::Account(old_primary_key), perms);
+                    record.add_secondary_keys(iter::once(sk.clone()));
+                    Self::deposit_event(RawEvent::SecondaryKeysAdded(target_did, vec![sk.into()]));
+                    Ok(())
+                })
```

Comment:
If optional permissions were passed to `common_rotate_primary_key`, then it could handle the two cases of either unlink the old primary key (`None`) or add the old key as a secondary key with permissions (`Some(Permissions)`).

This would also remove the extra get/mutate of `DidRecords`.

Reply:
I'd argue it's more readable as is, however am I right in assuming that the extra mutate will result in higher transaction costs? If that is the case then I'll definitely move it into the common function.

Reply:
The current code will increase the transaction cost.  Normally I would agree to keep the code as readable as possible, but we have to think about transaction costs.  Also please note that `DidRecord` (see type `Identity` in `primitives/src/identity.rs`) holds a vec of the secondary keys and an identity can have 2024 secondary keys.  Each load/store from storage of that value must encode/decode the whole value.

For this extrinsic the `DidRecords` is being loaded 3 times and stored twice.
1. Line 293 to get the old primary key.
2. In `common_rotate_primary_key` line 204: `let mut record = <DidRecords<T>>::get(target_did);`
3. Line 300.  The mutate must read the current value before calling the mutate closure.

It is possible to get that down to just one load and one store by doing all of the work in the `common_rotate_primary_key`.

For the other extrinsic there is only 2 loads and one store of the `DidRecords`.

Storage loads/stores should be minimised as much as possible.

Reply:
Agreed, I already pushed the change with minimized loads and stores